tokio-threadpool/src/lib.rs

```diff
@@ -1,117 +1,117 @@
-// not, a new worker thread is spawned.
-//
-// # Spawning futures
-//
-// The spawning behavior depends on whether a future was spawned from within a
-// worker or thread or if it was spawned from an external handle.
-//
-// When spawning a future while external to the thread pool, the current
-// strategy is to randomly pick a worker to submit the task to. The task is then
-// pushed onto that worker's mpsc channel.
-//
-// When spawning a future while on a worker thread, the task is pushed onto the
-// back of the current worker's deque.
-//
-// # Sleeping workers
-//
-// Sleeping workers are tracked using a treiber stack [1]. This results in the
-// thread that most recently went to sleep getting woken up first. When the pool
-// is not under load, this helps threads shutdown faster.
-//
-// Sleeping is done by using `tokio_executor::Park` implementations. This allows
-// the user of the thread pool to customize the work that is performed to sleep.
-// This is how injecting timers and other functionality into the thread pool is
-// done.
-//
-// [1]: https://en.wikipedia.org/wiki/Treiber_Stack
-//
-// # Notifying workers
-//
-// When there is work to be done, workers must be notified. However, notifying a
-// worker requires cross thread coordination. Ideally, a worker would only be
-// notified when it is sleeping, but there is no way to know if a worker is
-// sleeping without cross thread communication.
-//
-// The two cases when a worker might need to be notified are:
-//
-// 1) A task is externally submitted to a worker via the mpsc channel.
-// 2) A worker has a back log of work and needs other workers to steal from it.
-//
-// In the first case, the worker will always be notified. However, it could be
-// possible to avoid the notification if the mpsc channel has two or greater
-// number of tasks *after* the task is submitted. In this case, we are able to
-// assume that the worker has previously been notified.
-//
-// The second case is trickier. Currently, whenever a worker spawns a new future
-// (pushing it onto its deque) and when it pops a future from its mpsc, it tries
-// to notify a sleeping worker to wake up and start stealing. This is a lot of
-// notification and it **might** be possible to reduce it.
-//
-// Also, whenever a worker is woken up via a signal and it does find work, it,
-// in turn, will try to wake up a new worker.
-//
-// # `blocking`
-//
-// The strategy for handling blocking closures is to hand off the worker to a
-// new thread. This implies handing off the `deque` and `mpsc`. Once this is
-// done, the new thread continues to process the work queue and the original
-// thread is able to block. Once it finishes processing the blocking future, the
-// thread has no additional work and is inserted into the backup pool. This
-// makes it available to other workers that encounter a `blocking` call.
+//! A work-stealing based thread pool for executing futures.
```

Comment:
This documentation was mostly intended for those who are exploring the source. Do you think it is useful to users of the library?

Reply:
Yes! Most importantly, because this is the only place where the two types of threads (*worker* and *backup*) are described. These details are essential to gaining sufficient understanding of the `blocking` strategy.

Toward making these docs more generally useful, maybe the ¶ *Crate layout* section could be renamed ¶ *Implementation notes* and moved to the end?

Reply:
I would suggest that the three following sections stay as code comments as I don't see them providing any insight to users:

* Crate layout
* Sleeping workers
* Notifying workers

The other sections are probably fine to include in the docs.

Reply:
One point of potential clarification for backup threads. Is it correct to say that if there is a fully sleeping backup thread at the point where a new worker is needed (`blocking` called on an existing worker thread), then sleeping backup can be promoted to a new worker? Or is that more a potential future optimization then a current one?  It would also be fare if you told me that the rustdoc (-threadpool root) can omit that level of detail.

Reply:
As I also stated below, I don't think this an essential clarification needed to merge this.  Can improve that latter, with a more targetted/ less "general" PR.